pkg/virt-controller/services/template.go

```diff
@@ -1056,26 +1063,33 @@ func getNetworkToResourceMap(virtClient kubecli.KubevirtClient, vmi *v1.VirtualM
 	return
 }
 
-func getCniInterfaceList(vmi *v1.VirtualMachineInstance) (ifaceListString string, cniAnnotation string) {
+func getCniAnnotations(vmi *v1.VirtualMachineInstance) (cniAnnotations map[string]string, err error) {
 	ifaceList := make([]string, 0)
+	ifaceListMap := make([]map[string]string, 0)
+	cniAnnotations = make(map[string]string, 0)
 
-	for _, network := range vmi.Spec.Networks {
-		// set the type for the first network
-		// all other networks must have same type
+	for idx, network := range vmi.Spec.Networks {
+		// Set the type for the first network. All other networks must have same type.
 		if network.Multus != nil {
-			ifaceList = append(ifaceList, network.Multus.NetworkName)
-			if cniAnnotation == "" {
-				cniAnnotation = "k8s.v1.cni.cncf.io/networks"
+			ifaceMap := map[string]string{
+				"name":      network.Multus.NetworkName,
+				"interface": fmt.Sprintf("net%d", idx+1),
 			}
+			ifaceListMap = append(ifaceListMap, ifaceMap)
+			// We have to handle Genie separately because it doesn't support JSON format.
 		} else if network.Genie != nil {
 			ifaceList = append(ifaceList, network.Genie.NetworkName)
-			if cniAnnotation == "" {
-				cniAnnotation = "cni"
-			}
 		}
 	}
-
-	ifaceListString = strings.Join(ifaceList, ",")
+	if len(ifaceListMap) > 0 {
+		ifaceJsonString, err := json.Marshal(ifaceListMap)
+		if err != nil {
+			return map[string]string{}, fmt.Errorf("Failed to create JSON list from CNI interface map %s", ifaceListMap)
+		}
+		cniAnnotations[MultusNetworksAnnotation] = fmt.Sprintf("%s", ifaceJsonString)
+	} else if len(ifaceList) > 0 {
```

Comment:
perhaps we should be totally agnostic to the fact that two plugins are mutually exclusive and just rely on whether the list is empty or not (meaning, move the `if` check from under the `else` branch of the first check). But since in reality they are forbidden to be used for the same VMI on API level, I guess it's a minor concern.

Reply:
I prefer the current code, where it is clearer that it's either a coma-separated list or a json list, never both.

Reply:
@dankenigsberg it's a minor thing but I feel like the fact that these two plugins are mutually exclusive should not be explicitly captured that deep in the call stack and instead governed (as it is right now) on API admission only. If for some reason admission allowed both through then the code would assume that plugins are no longer exclusive. Template code is to produce pod spec spec that reflects VMI spec, not to enforce API payload validity.

This is all sophistry though.